model.py

```diff
@@ -26,19 +27,45 @@ def getScore(doc: Document) -> float:
     def step(doc: Document) -> None:
         scoreBefore = Model.getScore(doc)
 
-        actionList = []
+        # TODO make it a class not a tuple
+        actionList: list[tuple[Agent, Action, AgentView]] = []
         for agent in doc.agents:
             agentView = AgentView(doc, agent.x, agent.y)
-            action = agent.selectAction(agentView)
-            actionList.append((agent, action))
-            #print(agentView.ToPrettyString())
-            #applied = doc.applyActionToAgent(agent, action)
-            #print("action type: ", action.type, " success: ", applied)
+            actions: list[Action] = agent.selectActions(agentView)
+            sharedAct = Model.J(actions, agent, agentView, doc)
+            actionList.append((agent, sharedAct, agentView))
+
         doc.applyActionList(actionList)
 
         scoreAfter = Model.getScore(doc)
         reward = scoreAfter - scoreBefore
         print("reward: ", reward, " score: ", scoreAfter)
 
         for agent in doc.agents:
-            agent.L(reward)
+            agent.L(actionList, reward)
+
+        for agent in doc.agents:
+            agent.Q(actionList, reward)
+
+
+    @staticmethod
+    def J(actions: list[Action], thisAgent: Agent, agentView: AgentView, doc: Document) -> Action:
+        if len(actions) == 0:
+            return None
+        bids: list[list[float]] = []
+        for actId, action in enumerate(actions):
+            bids.append([])
+            for agent in doc.agents:
+                if thisAgent.distanceTo(agent) <= Config.agent_coms_distance:
+                    agentBid = agent.bid(action, agentView)
```

Comment:
maybe other agents should bid on the whole `actions` list selected by the acting agent, instead of each individual action - so agents can assign bids based on the selection of available actions.

Reply:
idk, hard to say what impact it will make at this point. It can help them, but if they have a good bidding model, they should be able to make this work anyway